File: ethereumj-core/src/test/java/org/ethereum/net/rlpx/discover/QueueTest.java
/*
 * The MIT License (MIT)
 *
 * Copyright 2017 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 * Copyright (c) [2016] [ <ether.camp> ]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package org.ethereum.net.rlpx.discover;

import org.junit.Test;

import java.util.Comparator;
import java.util.concurrent.TimeUnit;

/**
 * Created by Anton Nashatyrev on 03.08.2015.
 */
public class QueueTest {

    private boolean exception = false;

    @Test
    public void simple() throws Exception {
        final PeerConnectionTester.MutablePriorityQueue<String, String> queue =
                new PeerConnectionTester.MutablePriorityQueue<>(new Comparator<String>() {
            @Override
            public int compare(final String o1, final String o2) {
                return o1.compareTo(o2);
            }
        });

        final int threadCnt = 8;
        final int elemCnt = 1000;

        final Runnable adder = new Runnable() {
            @Override
            public void run() {
                try {
                    System.out.println("Adding...");
                    for (int i = 0; i < elemCnt && !exception; i++) {
                        queue.add("aaa" + i);
                        if (i % 100 == 0) Thread.sleep(10);
                    }
                    System.out.println("Done.");
                } catch (final Exception e) {
                    exception = true;
                    e.printStackTrace();
                }
            }
        };

        final ThreadGroup tg = new ThreadGroup("test");

        final Thread[] t1 = new Thread[threadCnt];

        for (int i = 0; i < t1.length; i++) {
            t1[i] = new Thread(tg, adder);
            t1[i].start();
        }


        final Runnable taker = new Runnable() {
            @Override
            public void run() {
                try {
                    System.out.println("Taking...");
                    for (int i = 0; i < elemCnt && !exception; i++) {
                        queue.poll(1, TimeUnit.SECONDS);
                    }
                    System.out.println("OK: " + queue.size());
                } catch (final Exception e) {
                    exception = true;
                    e.printStackTrace();
                }
            }
        };
        final Thread[] t2 = new Thread[threadCnt];

        for (int i = 0; i < t2.length; i++) {
            t2[i] = new Thread(tg, taker);
            t2[i].start();
        }

        for (final Thread thread : t1) {
            thread.join();
        }
        for (final Thread thread : t2) {
            thread.join();
        }

        if (exception) throw new RuntimeException("Test failed");
    }
}

File: ethereumj-core/src/test/java/org/ethereum/net/rlpx/discover/QueueTest.kt
/*
 * The MIT License (MIT)
 *
 * Copyright 2017 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 * Copyright (c) [2016] [ <ether.camp> ]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package org.ethereum.net.rlpx.discover

import org.junit.Test
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Created by Anton Nashatyrev on 03.08.2015.
 */
class QueueTest {

    private var exception = false

    @Test
    @Throws(Exception::class)
    fun simple() {
        val queue = PeerConnectionTester.MutablePriorityQueue(Comparator<String> { o1, o2 -> o1.compareTo(o2) })

        val threadCnt = 8
        val elemCnt = 1000

        val adder = Runnable {
            try {
                println("Adding...")
                var i = 0
                while (i < elemCnt && !exception) {
                    queue.add("aaa" + i)
                    if (i % 100 == 0) Thread.sleep(10)
                    i++
                }
                println("Done.")
            } catch (e: Exception) {
                exception = true
                e.printStackTrace()
            }
        }

        val tg = ThreadGroup("test")

        val t1 = arrayOfNulls<Thread>(threadCnt)

        for (i in t1.indices) {
            t1[i] = Thread(tg, adder)
            t1[i]?.start()
        }


        val taker = Runnable {
            try {
                println("Taking...")
                var i = 0
                while (i < elemCnt && !exception) {
                    queue.poll(1, TimeUnit.SECONDS)
                    i++
                }
                println("OK: " + queue.size)
            } catch (e: Exception) {
                exception = true
                e.printStackTrace()
            }
        }
        val t2 = arrayOfNulls<Thread>(threadCnt)

        for (i in t2.indices) {
            t2[i] = Thread(tg, taker)
            t2[i]?.start()
        }

        for (thread in t1) {
            thread?.join()
        }
        for (thread in t2) {
            thread?.join()
        }

        if (exception) throw RuntimeException("Test failed")
    }
}
